Source/GTMSessionFetcherService.m

```diff
@@ -224,6 +231,54 @@ - (GTMSessionFetcher *)fetcherWithURLString:(NSString *)requestURLString {
   return [self fetcherWithURL:url];
 }
 
+- (void)addHeaderDecorator:(id<GTMSessionFetcherHeaderDecorator>)decorator {
+  @synchronized(self) {
+    [_headerDecorators addPointer:(__bridge void *)decorator];
+  }
+}
+
+- (void)removeHeaderDecorator:(id<GTMSessionFetcherHeaderDecorator>)decorator {
+  @synchronized(self) {
+    NSUInteger i = 0;
+    for (id<GTMSessionFetcherHeaderDecorator> headerDecorator in _headerDecorators) {
+      if (headerDecorator == decorator) {
+        break;
+      }
+      ++i;
+    }
+    GTMSESSION_ASSERT_DEBUG(i < _headerDecorators.count, @"header decorator %@ must be passed to -addHeaderDecorator: before removing", decorator);
+    if (i < _headerDecorators.count) {
+      [_headerDecorators removePointerAtIndex:i];
+    }
+  }
+}
+
+- (void)decorateHeadersForRequest:(NSURLRequest *)request
+                          fetcher:(GTMSessionFetcher *)fetcher {
+  @synchronized(self) {
+    BOOL needsCompact = NO;
+    for (id<GTMSessionFetcherHeaderDecorator> headerDecorator in _headerDecorators) {
+      if (!headerDecorator) {
+        // The last reference to a weakly-held header decorator was released.  Explicitly compact
+        // the array after this is finished.
+        needsCompact = YES;
+        continue;
+      }
+      NSDictionary<NSString *, NSString *> *_Nullable headers = [headerDecorator additionalHeadersForRequest:request];
+      for (NSString *headerName in headers) {
+        NSString *headerValue = headers[headerName];
+        [fetcher setRequestValue:headerValue forHTTPHeaderField:headerName];
+      }
+    }
+    if (needsCompact) {
```

Comment:
I'm curious how much of a concern this is likely to be. I would imagine in most cases the decorator is intended to live the entire duration of the service, and the reason for using weak pointers is (I think) largely to avoid a retain cycle.

I would foresee most fetcher service instances will have 0, 1, or 2 decorators, and I imagine most decorators will have lifetimes largely the same as the service itself.

Reply:
I removed the compact logic to simplify it. If folks are adding hundreds of decorators and letting the weak references go nil, we can revisit this.